gitzones/auth

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/python3
+# this file is formatted using black
+import argparse
+import os
+import sys
+from typing import List, Dict
+
+authorized_keys_file = "/home/git/.ssh/authorized_keys"
+host_public_key_file = "/etc/ssh/keys/ssh_host_ed25519_key.pub"
+usage = """auth <command> [<args>]
+    
+Available commands are:
+  id                          Shows server key formatted for client's ~/.ssh/known_hosts
+  add <proto> <key> <label>   Adds a client key to the list of authorized keys.
+                              <proto> <key> <label> is typically the content of the public key file of the client.
+                              The label must be unique; usage of existing labels will overwrite the existing key
+                              without warning.
+  rm <label>                  Removes client key with given label from the list of authorized keys
+  ls [-v]                     Lists all authorized clients.
+  cp                          Print contents of the authorized_keys file.
+"""
+
+
+class AuthorizedKeysException(Exception):
+    pass
+
+
+class AuthorizedKeys(list):
+    """ Provides management of an SSH "authorized_keys" file, restricted to a subset of functionality. """
+
+    def __init__(self, location: str = authorized_keys_file) -> None:
+        super().__init__()
+        self.location = location
+        self.db = {}
+        try:
+            with open(location) as f:
+                for line in f:
+                    if not line.strip() or line.startswith("#"):
+                        continue
+                    proto, key, label = line.strip().split(" ", maxsplit=2)
+                    self.db[label] = proto + " " + key
+        except FileNotFoundError:
+            pass
+
+    def add(self, proto: str, key: str, label: str) -> None:
+        """ Adds an authorized key. """
+        if label in self.db:
+            raise AuthorizedKeysException(f'Key with label "{label}" already exists.')
+        self.db[label] = proto + " " + key
+        self._save()
+
+    def rm(self, label: str) -> None:
+        """ Removes an authorized key, identified by its label. Raises if key with given label cannot be found. """
+        try:
+            del self.db[label]
+        except KeyError:
+            raise AuthorizedKeysException(
+                f'Could not find authorized key with label "{label}".'
+            )
+        self._save()
+
+    def ls(self) -> Dict[str, str]:
+        """ Returns dictionary of authorized keys, identified by their labels. """
+        return self.db.copy()
+
+    def _save(self) -> None:
+        with open(self.location, "w") as f:
```

Comment:
As this is used in production and logins may be happening while the file is written, better write to a temporary file and then move it into place. This also prevents issues from error conditions occurring during write.

Reply:
done